libraries/node-core-library/src/FileSystem.ts

```diff
@@ -0,0 +1,432 @@
+   * Defaults to `"utf8"`.
+   */
+  encoding?: Encoding;
+
+  /**
+   * If specified, will normalize line endings to the specified style of newline.
+   * Defaults to `NewlineConversion.None`.
+   */
+  convertLineEndings?: NewlineConversion;
+}
+
+/**
+ * The options for FileSystem.move()
+ * @public
+ */
+export interface IMoveOptions {
+  /**
+   * If true, will overwrite the file if it already exists. Defaults to false.
+   */
+  overwrite?: boolean;
+
+  /**
+   * If true, will ensure the folder is created before writing the file.
+   * Defaults to `false`.
+   */
+  ensureFolder?: boolean;
+}
+
+/**
+ * The options for FileSystem.deleteFile()
+ * @public
+*/
+export interface IDeleteFileOptions {
+  /**
+   * If true, will throw an exception if the file did not exist before `deleteFile()` was called.
+   * Defaults to `false`.
+   */
+  throwIfNotExists?: boolean;
+}
+
+/**
+ * A standardized file system API. Mostly contains wrappers around `fs` and `fs-extra`.
+ * Also contains useful options to simplify and normalize file system operations.
+ * @public
+ */
+export class FileSystem {
+
+  // ===============
+  // COMMON OPERATIONS
+  // ===============
+
+  /**
+   * Returns true if the path exists on disk. It can be a file, folder, or link.
+   * Behind the scenes it uses `fs.existsSync()`.
+   * @param path - The absolute or relative path to the file, folder, or link.
+   */
+  public static exists(path: string): boolean {
+    return fsx.existsSync(path);
+  }
+
+  /**
+   * Gets the statistics for a particular file or folder.
+   * If the path is a link, this function follows the link and returns statistics about the link target.
+   * Behind the scenes it uses `fs.statSync()`.
+   * @param path - The absolute or relative path to the file, folder, or link.
+   */
+  public static getStatistics(path: string): fs.Stats {
+    return fsx.statSync(path);
+  }
+
+  /**
+   * Updates the accessed and modified timestamps of the file, folder, or link referenced by path.
+   * Behind the scenes it uses `fs.utimesSync()`.
+   * @param path - The path of the file that should be modified.
+   * @param accessedTime - The UNIX epoch time when this was last accessed.
+   * @param modifiedTime - The UNIX epoch time when this was last modified.
+   */
+  public static updateTimes(path: string, accessedTime: number, modifiedTime: number): void {
+    fsx.utimes(path, accessedTime, modifiedTime);
+  }
+
+  /**
+   * Changes the permissions mode of a file, folder, or link.
+   * Behind the scenes it uses `fs.chmodSync()`.
+   * @param path - The absolute or relative path to the object that should be updated.
+   * @param mode - should be a UNIX-style permissions modifier number (e.g. 777 or 666 etc)
+   */
+  public static changeMode(path: string, mode: number): void {
+    fsx.chmodSync(path, mode);
+  }
+
+  /**
+   * Moves a file. The folder must exist, unless the `ensureFolder` option is provided.
+   * Behind the scenes it uses `fsx.moveSync()`
```

Comment:
>fsx [](start = 32, length = 3)

Do we want to use `fsx` or `fs` internally? Seems odd to write a wrapper around a wrapper.

Reply:
I don't really think of `fsx` as a wrapper, more of a "here are some extra functions", since all it really does is augment `fs`.. Nonetheless, nothing from `fsx` is exposed, so we may be able to eliminate it in the future.

---
In reply to: [198670755](https://github.com/Microsoft/web-build-tools/pull/727#discussion_r198670755) [](ancestors = 198670755)